src/backend/distributed/master/master_truncate.c

```diff
@@ -65,22 +67,79 @@ citus_truncate_trigger(PG_FUNCTION_ARGS)
 
 	if (partitionMethod == DISTRIBUTE_BY_APPEND)
 	{
+		Oid schemaId = get_rel_namespace(relationId);
+		char *schemaName = get_namespace_name(schemaId);
+		char *relationName = get_rel_name(relationId);
+
 		DirectFunctionCall3(master_drop_all_shards,
 							ObjectIdGetDatum(relationId),
 							CStringGetTextDatum(relationName),
 							CStringGetTextDatum(schemaName));
 	}
 	else
 	{
-		StringInfo truncateStatement = makeStringInfo();
-		char *qualifiedTableName = quote_qualified_identifier(schemaName, relationName);
+		List *taskList = TruncateTaskList(relationId);
 
-		appendStringInfo(truncateStatement, "TRUNCATE TABLE %s CASCADE",
-						 qualifiedTableName);
-
-		DirectFunctionCall1(master_modify_multiple_shards,
-							CStringGetTextDatum(truncateStatement->data));
+		if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+		{
```

Comment:
Just brainstorming on this: I think one interesting note here is that with trigger approach we're *in a sense* mixing planning and execution.  `TRUNCATE parent_table CASCADE`, would first create list of tasks, then execute it for the parent table, then create list of tasks for child_1 -> execute it, create list of tasks for child_n -> execute it.

I think for `TRUNCATE` it doesn't matter because we're always acquiring access exclusive lock on the table. Anything interesting to you here?

Reply:
I think the current implementation is correct, but very slow. We don't really want to cascade on the coordinator if we are already going to cascade on the workers.